lib/streamlit/commands/page_config.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2018-2020 Streamlit Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from streamlit.delta_generator import _enqueue_message
+from streamlit.proto import ForwardMsg_pb2
+from streamlit.proto import PageConfig_pb2
+from streamlit.elements import image_proto
+from streamlit.errors import StreamlitAPIException
+
+
+def set_page_config(
+    page_title=None, page_icon=None, layout="centered", initial_sidebar_state="auto"
+):
+    """
+    Configures the default settings of the page.
+
+    .. note::
+        This must be the first Streamlit command used in your app, and must only
```

Comment:
Is this truly a "must", or a suggestion so that the page doesn't dance around? If it's the latter, it feels like we should put "it is suggested that this is the first command"

Reply:
Good question -- it actually is a must, as in "Once you've called st.title("blah"), calling st.set_page_config will be ignored and we'll show an exception"

The rationale behind forcing it to show first, and also exactly once, but still keep it separate from config options is described by the product team here:

https://www.notion.so/streamlit/Product-1472107ee200491b82894fe5fb5e4559?p=44a10601796749f6bf8c130c4f2892cc

I personally don't feel strongly about forcing it to be set once and at the top, but I do think it's nice to group these different settings, and also separate them from server-side options. That said, I agree that the overlap is a bit confusing and would welcome any suggestions to make the distinction clearer (expand the docs? rename the method?)

Reply:
> Once you've called st.title("blah"), calling st.set_page_config will be ignored and we'll show an exception instead.

This feels too stateful to me. I can imagine a series of questions that are effectively answered with "You have a call before `st.beta_set_page_configuration`. 

The idea that it's immutable is also strange to me. Does this mean each time the app is started, or each time the page is refreshed? If it's the former, then having to restart your app to change your page title will feel really heavy.

Reply:
I think Thiago and Adrien's positions were that it's easier to enforce "first and only once" and then relax that later, than to break some apps if we later decide "first and only once" is important. Basically if users bug us with examples of when they want to make calls before st.set_page_config we can change our minds more easily.

Personally I do think this enforcement is a weird pattern compared to our other APIs, and would be happy making both of these suggestions rather than enforcing it.

(I'm using "immutable" here really just to mean "should be called once" aka "second call will throw StreamlitAPIException"; it is changed each time the app is refreshed just like most streamlit calls.)

Reply:
@tvst Since this comes down to a product question, could you weigh in here?

Reply:
> I can imagine a series of questions that are effectively answered with "You have a call before st.beta_set_page_configuration".

Great point. We should show a clear error to the user when we they do this.
(BTW I _haven't_ looked through the PR to see if it already does this!)

As for the reasoning behind immutability, see the [spec discussion in Notion](https://www.notion.so/streamlit/Title-favicon-wide-mode-sidebar-mode-mega-proposal-44a10601796749f6bf8c130c4f2892cc).

The TLDR is that mutability comes with several problems, and we opted for the safe approach that allows us to make changes later. And, BTW, this is a beta feature, meaning its API may change depending on what we hear from the community!